File: solutions/0098-validate-binary-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool checkValidBST(TreeNode* root, TreeNode* low, TreeNode* high) {
        if (!root) {
            return true;
        }
        if ((low && root->val <= low->val) ||
            (high && root->val >= high->val)) {
            return false;
        }
        return checkValidBST(root->right, root, high) &&
               checkValidBST(root->left, low, root);
    }
    bool isValidBST(TreeNode* root) {
        return checkValidBST(root, nullptr, nullptr);
    }
};

File: solutions/0098-validate-binary-search-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func checkValidBST(root, low, high *TreeNode) bool {
    if root == nil {
        return true
    }
    // fmt.Print(root.Val)
    // fmt.Print("  ")
    if low != nil {
        // fmt.Print(", low: ")
        // fmt.Print(low.Val)
    }
    if high != nil {
        // fmt.Print(", high: ")
        // fmt.Print(high.Val)
    }
    if (low != nil && root.Val <= low.Val) ||
        (high != nil && root.Val >= high.Val) {
        return false
    }
    // fmt.Println()

    return checkValidBST(root.Left, low, root) &&
            checkValidBST(root.Right, root, high)
}

func isValidBST(root *TreeNode) bool {
    if root == nil ||
        (root.Left == nil && root.Right == nil) {
        return true
    }
    return checkValidBST(root, nil, nil)
}
